Elevator and Turret

teodortopan opened the pull request:
Finalized elevator and turret code

Commit:
elevator prototype code

controller errors due to no ocean libs

Co-Authored-By: sgowda260 <116316950+sgowda260@users.noreply.github.com>

Commit:
updated elevator

Co-Authored-By: sgowda260 <116316950+sgowda260@users.noreply.github.com>

Commit:
added code for lower switch

Co-Authored-By: sgowda260 <116316950+sgowda260@users.noreply.github.com>
Co-Authored-By: Abhinav Daram <76536884+SavageCabbage360@users.noreply.github.com>

ASH-will-WIN committed:
Turret Initial Commit

- Created Turret class
- Created run command for turret
- Made it so turret can't go past limit switches

Commit:
Added PID Command

ASH-will-WIN committed:
Merge branch 'subsystem/elevator-turret' of https://github.com/Troy-Argonauts/2023RobotCode into subsystem/elevator-turret

Commit:
updated elevator

Added periodic, PID command, values of limit switches and encoder

Co-Authored-By: sgowda260 <116316950+sgowda260@users.noreply.github.com>

Commit:
added javadoc comments

added javadoc comments to code

Co-Authored-By: sgowda260 <116316950+sgowda260@users.noreply.github.com>

ASH-will-WIN committed:
Added JavaDocs and PID to Turret

- Added JavaDocs
- Added PID
- Added Periodic variables that hold sensor values

Co-Authored-By: TeoElRey <98327660+TeoElRey@users.noreply.github.com>
Co-Authored-By: Isaac <54126919+Aizakkuno@users.noreply.github.com>

ASH-will-WIN committed:
Merge branch 'subsystem/elevator-turret' of https://github.com/Troy-Argonauts/2023RobotCode into subsystem/elevator-turret

ASH-will-WIN committed:
Added authors to java docs

Added authors to java docs

Co-Authored-By: TeoElRey <98327660+TeoElRey@users.noreply.github.com>
Co-Authored-By: Isaac <54126919+Aizakkuno@users.noreply.github.com>

Commit:
finalized turret elevator code

finalized code

Co-Authored-By: sgowda260 <116316950+sgowda260@users.noreply.github.com>
Co-Authored-By: Ashwin Shrivastav <103900575+ASH-will-WIN@users.noreply.github.com>

akashsarada reviewed (changes requested):
Elevator looks good!
1. Need to move constants to the constants file (work with @SolidityContract or @akashsarada)

Turret is almost ready
1. Need to move constants to the constants file (work with @SolidityContract or @akashsarada)
2. Do we need methods to get Limit Switch States if the variables are accessible?

Commit:
updated turret and elevator constants

added new variables to constants file and moved all existing IDs there

Co-Authored-By: Akash Sarada <95705632+akashsarada@users.noreply.github.com>

Commit:
finalized turret and elevator

ASH-will-WIN committed:
Removed get() methods for limit switches

- got rid of get() methods for limit switches
- made limit switch variables public

kelmalon commented:
Elevator Manipulator needs to be added for intake passover system

kelmalon commented:
Absolute Encoder needs to be added to turret

SolidityContract committed:
Updated Pneumatics

Updated Pneumatics Code for manipulator

ASH-will-WIN committed:
Got rid of Potentiomenter

ASH-will-WIN committed:
Added Absolute Encoder

Updated from using potentiometer to an Absolute Encoder.

Co-Authored-By: rohangajulaghosh <40278130+SolidityContract@users.noreply.github.com>

SolidityContract committed:
Updated Elevator and Turret

ASH-will-WIN committed:
Changed from using Absolute Encoder to Built-In encoder on turret

Changed from using absolute encoder to using built in encoder.

SolidityContract committed:
Removed Manipulator Code

Removed Manipulator code from elevator branch

ArgonautsStation committed:
Testing Changes

Added ArgoController Class
Changed CAN ID's
Commented out limit switches, as they have not been mounted yet.
Added debugging data to SmartDashboard

Co-Authored-By: Akash Sarada <95705632+akashsarada@users.noreply.github.com>

Commit:
updated javadocs

SolidityContract committed:
Updated Javadoc

Co-Authored-By: TeoElRey <98327660+TeoElRey@users.noreply.github.com>

ASH-will-WIN committed:
Updated Javadocs

Updated Javadocs

ASH-will-WIN committed:
Updated Javadocs

Updated Javadocs

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Turret.java:
```diff
@@ -0,0 +1,99 @@
+     * These values will later be used to make sure turrets reach positions and stay within limits.
+     */
+    @Override
+    public void periodic(){
```
Please update encoders to save to VARIABLES here, that we then check in PID instead of getPosition().

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Turret.java:
```diff
@@ -0,0 +1,99 @@
+//    }
+
+    public double getTurretPosition(){
+        return (turretMotor.getEncoder().getPosition() * 42);
```
Replace getPosition() with variable

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Turret.java:
```diff
@@ -0,0 +1,99 @@
+            pid,
+            () -> getTurretPosition(),
+
+            () -> turretMotor.getEncoder().getPosition(),
```
Remove

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Turret.java:
```diff
@@ -0,0 +1,99 @@
+    public void turretPID(double setpoint){
+        new PIDCommand (
+            pid,
+            () -> getTurretPosition(),
```
Replace with variable updated in periodic

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Elevator.java:
```diff
@@ -0,0 +1,125 @@
+     * @return encoder position based on encoder values
+     */
+    public double getPosition() {
+        return (rightMotor.getEncoder().getPosition() + leftMotor.getEncoder().getPosition()) / (2 * Constants.Elevator.kEncoderGearboxScale);
```
Replace getPosition() with public variable updated in periodic

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Elevator.java:
```diff
@@ -0,0 +1,125 @@
+     * @param speed desired elevator extension or retraction speed
+     */
+    public void setPower(double speed, double nerf) {
+        leftMotor.set(speed * nerf);
```
Add limit switch code and comment out until mounted

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Elevator.java:
```diff
@@ -0,0 +1,125 @@
+     * These values will later be used to make sure elevator extends or retracts to optimal position.
+     */
+    @Override
+    public void periodic() {
```
Set variables for encoder values and set their values here. Then use the variables in PIDs instead of getPosition().

kelmalon reviewed (changes requested):
Please make these changes. If you have any questions please reach out to me, Akash, Chirag, or Rohan.

DarkMG1 commented on src/main/java/org/troyargonauts/subsystems/Elevator.java:
```diff
@@ -0,0 +1,125 @@
+     * 
+     * @param setpoint will be the desired extension point
+     */
+    public void elevatorPID(double setpoint) {
```
Set PIDController's setpoint within method

DarkMG1 replied:
Refer to line 175 in [Drivetrain.java](https://github.com/Troy-Argonauts/2023RobotCode/blob/feature/drivetrain-pid/src/main/java/org/troyargonauts/subsystems/DriveTrain.java) on feature/drivetrain-pid branch for an example.

DarkMG1 commented on src/main/java/org/troyargonauts/subsystems/Turret.java:
```diff
@@ -0,0 +1,99 @@
+     * Using a PID command, turret will rotate to a setpoint using the PID Controller. 
+     * @param setpoint Setpoint for the Turret
+     */
+    public void turretPID(double setpoint){
```
Set PIDController's setpoint within method

DarkMG1 replied:
Refer to line 175 in [Drivetrain.java](https://github.com/Troy-Argonauts/2023RobotCode/blob/feature/drivetrain-pid/src/main/java/org/troyargonauts/subsystems/DriveTrain.java) on feature/drivetrain-pid branch for an example.

kelmalon replied:
Remove "Using a PID command," at the beginning of the javadoc

ASH-will-WIN committed:
Updated Javadocs and Code

- Made getTurretPosition() a periodic variable
- Fixed JavaDocs
- Fixed PID

ASH-will-WIN committed:
Fixed Periodic

ASH-will-WIN committed:
Updated PID

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Turret.java:
```diff
@@ -53,43 +62,41 @@ public void setPower(double power, double nerf){
+
+    public boolean getRightLimitSwitchState() {
+        return rightLimitSwitch.isLimitSwitchEnabled();
+    }
```
Remove Methods, switches are public, we can access them anywhere.

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Turret.java:
```diff
@@ -28,22 +29,30 @@ public class Turret extends SubsystemBase {
      * This sets motor to given power. Will not allow turret to go past the two magnetic limit switches.
      * @param power Desired motor power.
-     *
+     * @param nerf Nerf of motor power.
```
Add that it is a percentage nerf

kelmalon reviewed (commented):
See Inline edits. Remove limit switch methods as they are redundant, limit switches are public can access anywhere.

ASH-will-WIN committed:
Removed getLimitSwitch methods

-  Removed Both getLimitSwitch methods.
- Fixed JavaDocs
- Updated Smart Dashboard with limit switch states

Commit:
updated elevator

updated javadocs, get position method, periodic, and PID command

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Elevator.java:
```diff
@@ -103,9 +105,6 @@ public void setPower(double speed, double nerf) {
      * Returns encoder position based on encoder values
      * @return encoder position based on encoder values
      */
```
Remove Javadoc

kelmalon commented on src/main/java/org/troyargonauts/subsystems/Elevator.java:
```diff
@@ -103,9 +105,6 @@ public void setPower(double speed, double nerf) {
-    }
 
     /**
      * Using a PID command, the elevator will shift to a given setpoint using the
```
Remove "Using a PID Command," at start of sentence

kelmalon reviewed (changes requested):
Just a couple more things!

Commit:
updated javadocs

cleaned up elevator javadocs

ASH-will-WIN committed:
Moved PID Nerf to Constants

ASH-will-WIN committed:
Merge branch 'subsystem/elevator-turret' of https://github.com/Troy-Argonauts/2023RobotCode into subsystem/elevator-turret

kelmalon reviewed (approved):
Looking good! Ready to merge!

ArgonautsStation committed:
Fixed Turret PID Error

Co-Authored-By: rlokesh07 <63570802+rlokesh07@users.noreply.github.com>
Co-Authored-By: rohangajulaghosh <40278130+SolidityContract@users.noreply.github.com>
Co-Authored-By: Ashwin Shrivastav <103900575+ASH-will-WIN@users.noreply.github.com>

ArgonautsStation committed:
Created reset functions

Co-Authored-By: rlokesh07 <63570802+rlokesh07@users.noreply.github.com>
Co-Authored-By: rohangajulaghosh <40278130+SolidityContract@users.noreply.github.com>
Co-Authored-By: Ashwin Shrivastav <103900575+ASH-will-WIN@users.noreply.github.com>

kelmalon reviewed (approved):
looking good

ASH-will-WIN committed:
returned PID on Turret

ASH-will-WIN committed:
Merge branch 'subsystem/elevator-turret' of https://github.com/Troy-Argonauts/2023RobotCode into subsystem/elevator-turret

ASH-will-WIN committed:
Removed encoder get methods and updated constants

kelmalon reviewed (approved):
Good Job!

SolidityContract committed:
Added java docs

- Added nerf constant
- Fixed and added Javadocs

Co-Authored-By: Ashwin Shrivastav <103900575+ASH-will-WIN@users.noreply.github.com>

ArgonautsStation committed:
Turret PID

SolidityContract committed:
Updated Elevator Code

Added Appropriate Motors
Updated Constants

SolidityContract committed:
Updated Elevator

Updated Constants
Added Setpojnts

DarkMG1 commented:
Merged manually.
